### PR TITLE
Don't include assert.h from public header file

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
## Impact
Minor

## Testing
Pass CI, please ignore the follow warning:
```
Error: /home/runner/work/incubator-nuttx/incubator-nuttx/nuttx/arch/arm/src/lpc17xx_40xx/lpc17_40_usbdev.c:2607:2: error: Mixed case identifier found
```

### CODE REVIEW

**Comment by xiaoxiang781216** [comment]
Depends on: https://github.com/apache/incubator-nuttx-apps/pull/722

**Comment by xiaoxiang781216** [comment]
Depends on: https://github.com/apache/incubator-nuttx-apps/pull/731

**Review by jerpelea — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
depends on https://github.com/apache/incubator-nuttx-apps/pull/733

File: arch/arm/src/sama5/sam_can.c [hunk 1]
```diff
@@ -1604,7 +1605,7 @@ static void can_interrupt(int irq, void *context, FAR void *arg)
  *      (1 x tCAN).
  *   2. Propagation segment (PROP_SEG):  This part of the bit time is used
  *      to compensate for the physical delay times within the network. It is
- *      twice the sum of the signal�s propagation time on the bus line, the
+ *      twice the sum of the signal�s propagation time on the bus line, the
```
Comment (masayuki2009) [inline arch/arm/src/sama5/sam_can.c:1608]: @xiaoxiang781216 
unnecessary change?

Comment (xiaoxiang781216) [reply arch/arm/src/sama5/sam_can.c]: Yes, I revert the change. Please review again.

**Comment by xiaoxiang781216** [comment]
@jerpelea could you merge apache/incubator-nuttx-apps#733? so I can restart CI to very this PR.

**Comment by xiaoxiang781216** [comment]
depends on: https://github.com/apache/incubator-nuttx-apps/pull/738

**Comment by masayuki2009** [comment]
@xiaoxiang781216 
Could you please rebase the branch with the master and fix the conflicts?

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216
> Could you please rebase the branch with the master and fix the conflicts?

Done.

**Comment by xiaoxiang781216** [comment]
depends on https://github.com/apache/incubator-nuttx-apps/pull/741

**Comment by masayuki2009** [comment]
@xiaoxiang781216 

Hmm, the buid still fails.

```
====================================================================================
Configuration/Tool: nucleo-g070rb/nsh,CONFIG_ARMV7M_TOOLCHAIN_GNU_EABIL
------------------------------------------------------------------------------------
  Cleaning...
  Configuring...
  Disabling CONFIG_ARMV6M_TOOLCHAIN_GNU_EABIL
  Enabling CONFIG_ARMV7M_TOOLCHAIN_GNU_EABIL
  Building NuttX...
chip/stm32_tim.c: In function 'stm32_tim_setclock':
Error: chip/stm32_tim.c:556:3: error: implicit declaration of function 'DEBUGASSERT' [-Werror=implicit-function-declaration]
  556 |   DEBUGASSERT(dev != NULL);
      |   ^~~~~~~~~~~
cc1: all warnings being treated as errors
make[1]: *** [Makefile:138: stm32_tim.o] Error 1
make[1]: Target 'libarch.a' not remade because of errors.
```

**Comment by xiaoxiang781216** [comment]
@masayuki2009 @jerpelea the change finally pass CI, please help me merge it.

**Comment by davids5** [comment]
@xiaoxiang781216 I can merge if others are off line. Apps PR fist then this?

**Comment by davids5** [comment]
Oh I see it was merge already

**Comment by davids5** [comment]
@xiaoxiang781216 I am doing a sync with upstream... Getting lots of  

` more undefined references to `DEBUGASSERT' follow`

Was the method to move the include to all the source files as apposed to a a common header file (like nsh.h" for all nsh)?

What was the way you tested this? 

```
find ../ | xargs grep -l -s DEBUGASSERT | xargs grep -L -E 'assert.h|cassert'
../apps/netutils/netlib/netlib_getroute.c
../apps/fsutils/ipcfg/ipcfg_text.c
../apps/industry/foc/fixed16/foc_velocity.c
../apps/industry/foc/float/foc_velocity.c
../apps/system/hex2bin/hex2mem_main.c
../apps/system/hex2bin/hex2bin_main.c
../apps/system/zmodem/host/nuttx/config.h
../apps/canutils/libcanard/Makefile
../apps/examples/ftpc/ftpc_main.c
../apps/examples/esp32_himem/esp32_himem_main.c
../apps/examples/hidkbd/hidkbd_main.c
../apps/nshlib/nsh_stdsession.c
../apps/nshlib/nsh_vars.c
../nuttx/fs/aio/aioc_contain.c
../nuttx/fs/aio/aio_queue.c
../nuttx/fs/littlefs/Make.defs
../nuttx/fs/hostfs/hostfs_rpmsg_server.c
../nuttx/fs/hostfs/hostfs_rpmsg.c
../nuttx/libs/libdsp/lib_transform.c
../nuttx/libs/libdsp/lib_transform_b16.c
../nuttx/libs/libdsp/lib_observer.c
../nuttx/libs/libdsp/lib_misc_b16.c
../nuttx/libs/libdsp/lib_motor.c
../nuttx/libs/libdsp/lib_pid.c
../nuttx/libs/libdsp/lib_motor_b16.c
../nuttx/libs/libdsp/lib_foc_b16.c
../nuttx/libs/libdsp/lib_pid_b16.c
../nuttx/libs/libdsp/lib_misc.c
../nuttx/libs/libdsp/lib_foc.c
../nuttx/graphics/nxglib/nxglib_copyrun.h
../nuttx/arch/renesas/src/common/up_createstack.c
../nuttx/arch/renesas/src/common/up_usestack.c
../nuttx/arch/renesas/src/sh1/sh1_serial.c
../nuttx/arch/renesas/src/sh1/sh1_sigdeliver.c
../nuttx/arch/renesas/src/rx65n/rx65n_dtc.c
../nuttx/arch/renesas/src/rx65n/rx65n_usbdev.c
../nuttx/arch/renesas/src/rx65n/rx65n_sbram.c
../nuttx/arch/renesas/src/m16c/m16c_serial.c
../nuttx/arch/renesas/src/m16c/m16c_sigdeliver.c
../nuttx/arch/risc-v/src/common/riscv_createstack.c
../nuttx/arch/risc-v/src/common/riscv_usestack.c
../nuttx/arch/misoc/src/common/misoc_serial.c
../nuttx/arch/misoc/src/lm32/lm32_usestack.c
../nuttx/arch/misoc/src/lm32/lm32_createstack.c
../nuttx/arch/misoc/src/lm32/lm32_irq.c
../nuttx/arch/misoc/src/minerva/minerva_createstack.c
../nuttx/arch/misoc/src/minerva/minerva_usestack.c
../nuttx/arch/misoc/src/minerva/minerva_irq.c
../nuttx/arch/z80/src/ez80/ez80_sigdeliver.c
../nuttx/arch/z80/src/ez80/ez80_serial.c
../nuttx/arch/z80/src/ez80/ez80_rtc.c
../nuttx/arch/z80/src/common/z80_usestack.c
../nuttx/arch/z80/src/common/z80_createstack.c
../nuttx/arch/z80/src/z80/z80_sigdeliver.c
../nuttx/arch/z80/src/z180/z180_mmu.c
../nuttx/arch/z80/src/z180/z180_sigdeliver.c
../nuttx/arch/z80/src/z8/z8_sigdeliver.c
../nuttx/arch/z80/src/z8/z8_serial.c
../nuttx/arch/sim/src/sim/up_qspiflash.c
../nuttx/arch/sim/src/sim/up_createstack.c
../nuttx/arch/sim/src/sim/up_usestack.c
../nuttx/arch/sim/src/sim/up_lcd.c
../nuttx/arch/mips/src/common/mips_createstack.c
../nuttx/arch/mips/src/common/mips_usestack.c
../nuttx/arch/z16/src/common/z16_createstack.c
../nuttx/arch/z16/src/common/z16_sigdeliver.c
../nuttx/arch/z16/src/common/z16_usestack.c
../nuttx/arch/z16/src/z16f/z16f_timerisr.c
../nuttx/arch/z16/src/z16f/z16f_serial.c
../nuttx/arch/x86_64/src/intel64/up_createstack.c
../nuttx/arch/x86_64/src/intel64/up_usestack.c
../nuttx/arch/x86/src/i486/up_createstack.c
../nuttx/arch/x86/src/i486/up_usestack.c
../nuttx/arch/arm/src/moxart/moxart_irq.c
../nuttx/arch/arm/src/cxd56xx/cxd56_hostif.c
../nuttx/arch/arm/src/cxd56xx/cxd56_powermgr_procfs.c
../nuttx/arch/arm/src/cxd56xx/cxd56_sph.c
../nuttx/arch/arm/src/cxd56xx/cxd56_i2c_bitbang.c
../nuttx/arch/arm/src/efm32/efm32_flash.c
../nuttx/arch/arm/src/efm32/efm32_serial.c
../nuttx/arch/arm/src/stm32h7/stm32_rtc.c
../nuttx/arch/arm/src/stm32f7/stm32_dumpgpio.c
../nuttx/arch/arm/src/stm32f7/stm32_rtc.c
../nuttx/arch/arm/src/stm32f7/stm32_dma2d.c
../nuttx/arch/arm/src/stm32f7/stm32_capture.c
../nuttx/arch/arm/src/tiva/common/tiva_dumpgpio.c
../nuttx/arch/arm/src/tiva/common/tiva_hciuart.c
../nuttx/arch/arm/src/lpc17xx_40xx/lpc17_40_lcd.c
../nuttx/arch/arm/src/lpc17xx_40xx/lpc17_40_allocateheap.c
../nuttx/arch/arm/src/am335x/am335x_can.c
../nuttx/arch/arm/src/lc823450/lc823450_timer.c
../nuttx/arch/arm/src/dm320/dm320_usbdev.c
../nuttx/arch/arm/src/a1x/a1x_irq.c
../nuttx/arch/arm/src/nuc1xx/nuc_dumpgpio.c
../nuttx/arch/arm/src/stm32l4/stm32l4_dumpgpio.c
../nuttx/arch/arm/src/stm32l4/stm32l4_usbdev.c
../nuttx/arch/arm/src/stm32l4/stm32l4_can.c
../nuttx/arch/arm/src/stm32l4/stm32l4_lptim.c
../nuttx/arch/arm/src/tms570/tms570_lowputc.c
../nuttx/arch/arm/src/stm32l5/stm32l5_tim.c
../nuttx/arch/arm/src/s32k1xx/s32k1xx_rtc.c
../nuttx/arch/arm/src/s32k1xx/s32k1xx_pindma.c
../nuttx/arch/arm/src/s32k1xx/s32k1xx_pindump.c
../nuttx/arch/or1k/src/mor1kx/up_irq.c
../nuttx/arch/or1k/src/common/up_usestack.c
../nuttx/arch/xtensa/src/esp32/esp32_wifi_utils.c
../nuttx/arch/xtensa/src/esp32s2/esp32s2_user.c
../nuttx/arch/hc/src/common/up_createstack.c
../nuttx/arch/hc/src/common/up_usestack.c
../nuttx/arch/hc/src/m9s12/m9s12_ethernet.c
../nuttx/arch/hc/src/m9s12/m9s12_serial.c
../nuttx/Documentation/contributing/coding_style.rst
../nuttx/net/ipforward/ipv4_forward.c
../nuttx/net/udp/udp_sendto_buffered.c
../nuttx/net/icmp/icmp_netpoll.c
../nuttx/net/icmpv6/icmpv6_rnotify.c
../nuttx/net/icmpv6/icmpv6_autoconfig.c
../nuttx/net/icmpv6/icmpv6_netpoll.c
../nuttx/net/icmpv6/icmpv6_notify.c
../nuttx/net/tcp/tcp_send_buffered.c
../nuttx/net/tcp/tcp_sendfile.c
../nuttx/drivers/motor/motor.c
../nuttx/drivers/input/nunchuck.c
../nuttx/drivers/leds/ncp5623c.c
../nuttx/drivers/leds/pca9635pw.c
../nuttx/drivers/power/bq2429x.c
../nuttx/drivers/eeprom/spi_xx25xx.c
../nuttx/drivers/eeprom/i2c_xx24xx.c
../nuttx/drivers/wireless/ieee802154/xbee/xbee.h
../nuttx/drivers/wireless/bluetooth/bt_uart_bcm4343x.c
../nuttx/drivers/analog/ads1242.c
../nuttx/drivers/syslog/syslog_rpmsg_server.c
../nuttx/drivers/i2c/i2c_bitbang.c
../nuttx/drivers/i2c/pca9540bdp.c
../nuttx/drivers/rf/dat-31r5-sp.c
../nuttx/sched/irq/irq_chain.c
../nuttx/sched/sched/sched_critmonitor.c
../nuttx/sched/clock/clock_time2ticks.c
../nuttx/mm/mm_heap/mm.h
../nuttx/mm/shm/shmat.c
../nuttx/mm/shm/shmdt.c
../nuttx/Kconfig
../nuttx/boards/risc-v/esp32c3/esp32c3-devkit/src/esp32c3_spiflash.c
../nuttx/boards/mips/pic32mx/sure-pic32mx/src/pic32mx_lcd1602.c
../nuttx/boards/mips/pic32mx/pic32mx7mmb/src/pic32_mio283qt2.c
../nuttx/boards/arm/cxd56xx/drivers/sensors/ak09912_scu.c
../nuttx/boards/arm/cxd56xx/drivers/sensors/apds9930_scu.c
../nuttx/boards/arm/cxd56xx/drivers/sensors/kx022_scu.c
../nuttx/boards/arm/cxd56xx/drivers/sensors/lt1pa01_scu.c
../nuttx/boards/arm/cxd56xx/drivers/sensors/bm1422gmv_scu.c
../nuttx/boards/arm/cxd56xx/drivers/sensors/bh1745nuc_scu.c
../nuttx/boards/arm/cxd56xx/drivers/sensors/rpr0521rs_scu.c
../nuttx/boards/arm/cxd56xx/drivers/sensors/bmp280_scu.c
../nuttx/boards/arm/cxd56xx/drivers/sensors/bh1721fvc_scu.c
../nuttx/boards/arm/cxd56xx/drivers/sensors/bmi160_scu.c
../nuttx/boards/arm/cxd56xx/drivers/sensors/bm1383glv_scu.c
../nuttx/boards/arm/rp2040/raspberrypi-pico/src/rp2040_gpio.c
../nuttx/boards/arm/imxrt/imxrt1060-evk/src/imxrt_ft5x06.c
../nuttx/boards/arm/imxrt/imxrt1064-evk/src/imxrt_ft5x06.c
../nuttx/boards/arm/imx6/sabre-6quad/README.txt
../nuttx/boards/arm/stm32/clicker2-stm32/src/stm32_automount.c
../nuttx/boards/arm/stm32/nucleo-f410rb/src/stm32_userleds.c
../nuttx/boards/arm/stm32/olimex-stm32-p407/src/stm32_djoystick.c
../nuttx/boards/arm/stm32/hymini-stm32v/src/stm32_r61505u.c
../nuttx/boards/arm/stm32/common/src/stm32_zerocross.c
../nuttx/boards/arm/stm32/nucleo-f446re/src/stm32_ajoystick.c
../nuttx/boards/arm/stm32/nucleo-f446re/src/stm32_userleds.c
../nuttx/boards/arm/stm32/stm3210e-eval/src/stm32_djoystick.c
../nuttx/boards/arm/stm32/nucleo-f4x1re/src/stm32_ajoystick.c
../nuttx/boards/arm/stm32/nucleo-f4x1re/src/stm32_userleds.c
../nuttx/boards/arm/stm32f0l0g0/nucleo-f091rc/src/stm32_userleds.c
../nuttx/boards/arm/stm32f0l0g0/nucleo-f072rb/src/stm32_userleds.c
../nuttx/boards/arm/samv7/samv71-xult/README.txt
../nuttx/boards/arm/stm32l4/stm32l4r9ai-disco/src/stm32_userleds.c
../nuttx/boards/arm/stm32l4/nucleo-l452re/src/stm32_userleds.c
../nuttx/boards/arm/stm32l4/stm32l476vg-disco/src/stm32_userleds.c
../nuttx/boards/arm/stm32l4/nucleo-l476rg/src/stm32_ajoystick.c
../nuttx/boards/arm/stm32l4/nucleo-l476rg/src/stm32_userleds.c
../nuttx/boards/arm/stm32l4/nucleo-l432kc/src/stm32_zerocross.c
../nuttx/boards/arm/stm32l4/nucleo-l432kc/src/stm32_userleds.c
../nuttx/boards/arm/lpc54xx/lpcxpresso-lpc54628/src/lpc54_ft5x06.c
../nuttx/boards/xtensa/esp32/esp32-ethernet-kit/src/esp32_spiflash.c
../nuttx/boards/xtensa/esp32/esp32-wrover-kit/src/esp32_spiflash.c
../nuttx/boards/xtensa/esp32/esp32-devkitc/src/esp32_spiflash.c
```

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216 I am doing a sync with upstream... Getting lots of
> 
> `more undefined references to`DEBUGASSERT' follow`
> 
> Was the method to move the include to all the source files as apposed to a a common header file (like nsh.h" for all nsh)?
> 

Yes, we can add assert.h into an internal header file to avoid the repeat inclusion, but this file shouldn't be inside nuttx/include.

> What was the way you tested this?
> 

Pass the CI, but some source files may be skipped as you found since they aren't enable to compiled by any defconfig.

> ```
> find ../ | xargs grep -l -s DEBUGASSERT | xargs grep -L -E 'assert.h|cassert'
> ../apps/netutils/netlib/netlib_getroute.c
> ../apps/fsutils/ipcfg/ipcfg_text.c
> ../apps/industry/foc/fixed16/foc_velocity.c
> ../apps/industry/foc/float/foc_velocity.c

I will create new patch to fix these files.

**Comment by davids5** [comment]
> > @xiaoxiang781216 I am doing a sync with upstream... Getting lots of
> > `more undefined references to`DEBUGASSERT' follow`
> > Was the method to move the include to all the source files as apposed to a a common header file (like nsh.h" for all nsh)?
> 
> Yes, we can add assert.h into an internal header file to avoid the repeat inclusion, but this file shouldn't be inside nuttx/include.
> 
> > What was the way you tested this?
> 
> Pass the CI, but some source files may be skipped as you found since they aren't enable to compiled by any defconfig.
> 
> > ```
> > find ../ | xargs grep -l -s DEBUGASSERT | xargs grep -L -E 'assert.h|cassert'
> > ../apps/netutils/netlib/netlib_getroute.c
> > ../apps/fsutils/ipcfg/ipcfg_text.c
> > ../apps/industry/foc/fixed16/foc_velocity.c
> > ../apps/industry/foc/float/foc_velocity.c
> > ```
> 
> I will create new patch to fix these files.


@xiaoxiang781216 - thanks - I had some WIP not listed above. 
Now in PR #3874 and https://github.com/apache/incubator-nuttx-apps/pull/753
